CI for ADB

dharmit opened the pull request:
CI tests for ADB to fix [#195](https://github.com/projectatomic/adb-atomic-developer-bundle/issues/195) on [projectatomic/adb-atomic-developer-bundle](https://github.com/projectatomic/adb-atomic-developer-bundle/).

More tests would be added to this in future. However, not much should change in the Jenkins configuration.

dustymabe commented:
I still think the run.py file should be very dumb similar to this [run.py](https://github.com/projectatomic/adb-tests/blob/master/cdrage-atomicapp-ci/centos-ci-jjb/run.py) where you just define a TEST_CMD environment var . Maybe consider putting the logic for running ansible elsewhere? 

one way to do this would be to have a seperate script that sets up ansible and then your TEST_CMD is defined inside your jjb file and is just `TEST_CMD="ansible_setup.sh && test_{provider}.sh"` or something like that.

dharmit commented:
@dustymabe `run.py` is doing more than just setting up the environment and executing the tests. It's actually figuring out which provider needs to be tested as well. Let's say you modify `components/centos/centos-k8s-singlenode-setup` or `components/centos/centos-openshift-setup`; it figures out which of the provider needs to be tested.

Having said that, I am open to ideas about how this can be done better. Currently it's hacky and doesn't look good enough to me. Ideally, having something like, a test that exclusively runs if custom Vagrantfile for Kubernetes is modified; another test exclusively for OpenShift Vagrantfile, and so on would make more sense to me.

However, I am not sure how to go about implementing that. Looking for ideas/suggestions.

dustymabe commented:
> @dustymabe run.py is doing more than just setting up the environment and executing the tests. It's actually figuring out which provider needs to be tested as well. Let's say you modify components/centos/centos-k8s-singlenode-setup or components/centos/centos-openshift-setup; it figures out which of the provider needs to be tested.

What I'm saying is that there is no reason why run.py (the jjb embedded python file) has to figure this out. Let run.py be dumb and just call out to another script (call it test-adb.py) that lives in the "adb-tests" repo. That script should be the one that has the complex logic in it. There are a few benefits to this:
- run.py is the same for all of our tests that run in the centos-ci
- you can easily run the tests on a machine that isn't the centos-ci (by calling test-adb.py with the right args)

dharmit commented:
@dustymabe Gotcha. I'll modify the PR and pingback soon,

dharmit commented:
While we're discussing this, I've seen an issue with the current approach. I'll try to explain it here.

Currently, if a PR is modifying two custom Vagrantfiles, let's say for OpenShift and k8s, the script detects that and runs tests for both Vagrantfiles one after the another. 

Let's say OpenShift Vagrantfile tests are getting executed first and then k8s. If any of the tests for former fail, and if none of the tests for latter fail, it is considered a passed overall test. Should we have separate tests for k8s and OpenShift providers? I think that's what you're doing for atomicapp. Having one Jenkins job trigger another would make sense here?

dkutalek commented:
I agree that tests should be instructed what to test (eg. by parameters)
as opposed to figure it out in tests. Not only the diagnostic could go wrong,
but software under test could be wrongly composed and test would
return wrongly pass instead of fail. 

Instead, test should just assert that environment is set up as required.

Example: when testing vagrant file for Openshift which actually start k8s instead.
Test should be instructed to test openshift and should fail on asserting it is running.

As to how many jobs per PR, I think one is enough.
Basic use case would be run all the tests for every PR,
running just some based on priority / resources / files changed 
is just optimization, isn't it? Then all tests should pass or be fixed.

With regards to test runner, why not use some existing one like eg. Avocado?
https://avocado-framework.readthedocs.org/en/latest/

It can run python, bash, whatever tests and report consistently.
It is being actively developed and used.

Thing is, we will need more features as we go and why wasting our resources
to develop another runner from scratch?

dustymabe commented:
> Let's say OpenShift Vagrantfile tests are getting executed first and then k8s. If any of the tests for former fail, and if none of the tests for latter fail, it is considered a passed overall test. Should we have separate tests for k8s and OpenShift providers? I think that's what you're doing for atomicapp. Having one Jenkins job trigger another would make sense here?

For atomicapp we run all of the tests every time so we don't have this problem. What you could do is have the test run every time but just make it exit success if there is nothing to test.

dharmit commented:
@dustymabe I've made modifications as you suggested. `run.py` is pretty much as dumb as you suggested it to be. Rest has been moved to individual test scripts.

dustymabe commented:
hey @dharmit.. do you mind rebasing this on master?

praveenkumar commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+
+url_base = "http://admin.ci.centos.org:8080"
+api_key = os.environ['API_KEY']
+count = os.environ['MACHINE_COUNT']
```
If you don't want to get keyerror in case 'API_KEY' not present then better use `os.getenv`
```
>>> os.getenv('HOME')
'/home/prkumar'
>>> os.getenv('HOM')
>>> os.environ['HOME']
'/home/prkumar'
>>> os.environ['HOM']
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
KeyError: 'HOM'
```

praveenkumar commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+    ssh_cmd = "ssh -t -t "
+    ssh_cmd += "-o UserKnownHostsFile=/dev/null "
+    ssh_cmd += "-o StrictHostKeyChecking=no "
+    ssh_cmd += "root@%s " % (host)
```
Better

```
ssh_cmd  = ("ssh -t -t ",
            "-o UserKnownHostsFile=/dev/null ",
            "-o StrictHostKeyChecking=no ",
            "root@%s ") % (host)
```

dharmit replied:
Done

navidshaikh commented on adb-vagrantfile-tests/openshift-tests/test-openshift.py:
```diff
@@ -0,0 +1,113 @@
+import subprocess
+import unittest
+import re
+import os
```
Not a blocker, try always to import the libs in the alphabetical order.

dharmit replied:
Done

praveenkumar commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+    remote_cmd += "cd adb-ci-ansible && ./install-ansible.sh && "
+    remote_cmd += "ansible-playbook install-adb.yml --extra-vars " \
+        "'adb_tests_dharmit=true clone_adb_repo=true'"
+
```
Same as above.

dharmit replied:
pyflakes complains about invalid syntax on the line `"git clone %s && " % ansible_repo_url` if I do the same as above:

~~~
remote_cmd = ("yum -y -q install git && "
              "git clone %s && " % ansible_repo_url
              "cd adb-ci-ansible && ./install-ansible.sh && "
              "ansible-playbook install-adb.yml --extra-vars "              
              "'adb_tests_dharmit=true clone_adb_repo=true'")
~~~

praveenkumar replied:
@dharmit you should do below to get rid of invalid syntax errror.

remote_cmd = ("yum -y -q install git && "
              "git clone %s && "
              "cd adb-ci-ansible && ./install-ansible.sh && "
              "ansible-playbook install-adb.yml --extra-vars "              
              "'adb_tests_dharmit=true clone_adb_repo=true'") % ansible_repo_url

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/README.md:
```diff
@@ -0,0 +1,33 @@
+### Current flow
+
+
```
In long term, it would be good to provide some context in README itself.

dharmit replied:
What would you like me to add? About what tests are being performed at the moment?

navidshaikh replied:
yeah, basically provide more detail about it. Not required ATM tough, but keep a note once we have minimal viable up.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/README.md:
```diff
@@ -0,0 +1,33 @@
+
+- Once Duffy provides a system for testing, `run.py` installs git and ansible
+  on it. Then the ansible playbook in *placeholder*
+  [ci-ansible](https://github.com/dharmit/ci-ansible) repo gets executed on the
```
Are we moving this repo as well from your namespace to projectatomic namespace?

dharmit replied:
That's the intention.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/README.md:
```diff
@@ -0,0 +1,33 @@
+  system which installs required packages on the system.
+
+- Ansible playbooks need to be moved to a proper place. However, they cannot be
+  written on per-test bassis either since many tests will require installing
```
*basis

dharmit replied:
Done

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/README.md:
```diff
@@ -0,0 +1,33 @@
+  instead.
+
+- Post playbook execution, we clone my fork of
+  [adb-tests](https://github.com/dharmit/adb-tests) repo and execute the tests.
```
Why does it clone your fork but not the original repo?

dharmit replied:
Because this PR is intended to put things into the original repo. If it gets an ack, I'll replace `dharmit` with `projectaotmic` and remove the references to this branch.

dharmit replied:
Although that would lead to a few iterations of testing if the tests are in place! :smile:

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/README.md:
Okay.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+test_cmd += "export ghprbPullId=%s && " % os.environ["ghprbPullId"]
+test_cmd += os.environ['TEST_CMD']
+
+jsondata = urllib.urlopen(req_url).read()
```
We'll also need some catch except here.

dharmit replied:
I'm okay to do that. However, the CentOS CI infra guys make sure that if environment is up, it's up as a whole and not in bits and pieces. That's been my experience so far. I can see the kind of situation we'd like to handle with `try..except`, but chances are it wouldn't come up. Anyway, point noted and implemented.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+        cmd = '%s "%s"' % (ssh_cmd, test_cmd)
+
+        print("Running cmd: {}".format(cmd))
+        exit_code = subprocess.call(cmd, shell=True)
```
try block will only run the commands, defining commands should be outside of try block

dharmit replied:
The `try..except` I added here was to make sure that even if my code's failing, the CentOS CI node that I've been allocated gets free. Should remove this from the final PR.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+        print("Running cmd: {}".format(cmd))
+        exit_code = subprocess.call(cmd, shell=True)
+    except:
+        pass
```
I think we should not just pass this, should we?

dharmit replied:
The `try..except` I added here was to make sure that even if my code's failing, the CentOS CI node that I've been allocated gets free. Should remove this from the final PR.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/test-k8s.py:
```diff
@@ -0,0 +1,148 @@
+    files = subprocess.check_output(['git', 'diff-tree', '--no-commit-id',
+                                     '--name-only', '-r', "%s" %
+                                     (os.environ['ghprbActualCommit'])],
+                                    cwd="/root/adb")
```
Nice

dharmit replied:
Thanks! :smile:

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/test-k8s.py:
```diff
@@ -0,0 +1,148 @@
+        except:
+            pass
+        self.assertIn('127.0.0.1', output.split())
+        self.assertIn('Ready', output.split())
```
I think, there should be better way to verify vagrant ssh

dharmit replied:
I think I should rename this something specific to Kubernetes check instead of checking Vagrant SSH. That's what it's doing anyway.

dharmit replied:
Renamed this to `test_03_kubectl_output` to make it more self-explanatory.

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/test-k8s.py:
```diff
@@ -0,0 +1,148 @@
+            ])
+            subprocess.call([
+                "vagrant", "ssh", "-c", "%s" %
+                "git clone https://github.com/dharmit/ci-ansible"
```
the ci-ansible repo should be defined in a constant

dharmit replied:
Done

navidshaikh commented on adb-vagrantfile-tests/k8s-tests/test-k8s.py:
```diff
@@ -0,0 +1,148 @@
+                "kubectl get pods| grep helloapache"
+            ])
+            self.assertIn("helloapache", output)
+            self.assertIn("Running", output)
```
and we can also write a test to curl at IP:PORT where helloapache is running

dharmit replied:
Leaving it for other contributors to add! :wink:

navidshaikh commented on adb-vagrantfile-tests/openshift-tests/README.md:
```diff
@@ -0,0 +1,33 @@
+  system which installs required packages on the system.
+
+- Ansible playbooks need to be moved to a proper place. However, they cannot be
+  written on per-test bassis either since many tests will require installing
```
*basis

navidshaikh commented on adb-vagrantfile-tests/openshift-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+    remote_cmd += "cd adb-ci-ansible && ./install-ansible.sh && "
+    remote_cmd += "ansible-playbook install-adb.yml --extra-vars " \
+        "'adb_tests_dharmit=true clone_adb_repo=true'"
+
```
as suggested above to define the strings

navidshaikh commented on adb-vagrantfile-tests/openshift-tests/run.py:
```diff
@@ -0,0 +1,60 @@
+        cmd = '%s "%s"' % (ssh_cmd, test_cmd)
+
+        print("Running cmd: {}".format(cmd))
+        exit_code = subprocess.call(cmd, shell=True)
```
and only commands should execute in try block, rest definition go out of the block

dharmit commented:
@navidshaikh @praveenkumar I've incorporated most of the changes that you guys have suggested. An important point to note that for tests to work after closing the PR, I'll have to change references to my fork and branch within to `projectatomic/adb-tests`. This can be done either before the merge or after - however you guys recommend. 

For both the providers - Kubernetes and OpenShift - there's not many tests at the moment. In fact I removed one test since last evening which was checking `kubectl` output in OpenShift box. This is obviously something we should expand upon.

dharmit commented:
To check if things are working, you can take a look at the fake PRs I've raised on ADB. [This for OpenShift](https://github.com/projectatomic/adb-atomic-developer-bundle/pull/358) and [this for Kubernetes](https://github.com/projectatomic/adb-atomic-developer-bundle/pull/317). The trigger phrases that work at the moment are `buildkubernetes`, `buildopenshift` and `buildall`.

navidshaikh commented:
> I'll have to change references to my fork and branch within to projectatomic/adb-tests. This can be done either before the merge or after - however you guys recommend.

I'd recommend you update the references now before merge.

dharmit commented:
@navidshaikh @praveenkumar I've updated the references. However, I'd be able to test them only after the PR's merged. And till it's merged, there might be few failing tests. :unamused:

navidshaikh commented:
LGTM!

dharmit commented:
Thanks, I'm checking if things are working well after removing the references to my forked repo.
